browser/net/brave_network_audit_browsertest.cc

```diff
@@ -0,0 +1,249 @@
+}
+
+class BraveNetworkAuditTest : public InProcessBrowserTest {
+ public:
+  BraveNetworkAuditTest() = default;
+
+  void SetUpCommandLine(base::CommandLine* command_line) override {
+    base::FilePath source_root_path;
+    base::PathService::Get(base::DIR_SOURCE_ROOT, &source_root_path);
+
+    // Full log containing all the network requests.
+    net_log_path_ = source_root_path.AppendASCII("network_log.json");
+
+    // Log containing the results of the audit only.
+    audit_results_path_ =
+        source_root_path.AppendASCII("network_audit_results.json");
+
+    command_line->AppendSwitchPath(network::switches::kLogNetLog,
+                                   net_log_path_);
+    command_line->AppendSwitchASCII(network::switches::kNetLogCaptureMode,
+                                    "Everything");
+  }
+
+  void TearDownInProcessBrowserTestFixture() override {
+    VerifyNetworkAuditLog();
+  }
+
+ private:
+  // Verify that the netlog file was written, appears to be well formed, and
+  // includes the requested level of data.
+  void VerifyNetworkAuditLog() {
+    // Read the netlog from disk.
+    std::string file_contents;
+    ASSERT_TRUE(base::ReadFileToString(net_log_path_, &file_contents))
+        << "Could not read: " << net_log_path_;
+
+    // Parse it as JSON.
+    auto parsed = base::JSONReader::ReadDeprecated(file_contents);
+    ASSERT_TRUE(parsed);
+
+    // Ensure the root value is a dictionary.
+    base::DictionaryValue* main;
+    ASSERT_TRUE(parsed->GetAsDictionary(&main));
+
+    // Ensure it has a "constants" property.
+    base::DictionaryValue* constants;
+    ASSERT_TRUE(main->GetDictionary("constants", &constants));
+    ASSERT_FALSE(constants->DictEmpty());
+
+    // Ensure it has an "events" property.
+    base::ListValue* events;
+    ASSERT_TRUE(main->GetList("events", &events));
+    ASSERT_FALSE(events->GetList().empty());
+
+    EXPECT_TRUE(PerformNetworkAuditProcess(events))
+        << "network-audit FAILED. Import " << net_log_path_.AsUTF8Unsafe()
+        << " in chrome://net-internals for more details.";
+
+    // Write results of the audit to disk, useful for further debugging.
+    WriteNetworkAuditResultsToDisk(*main, audit_results_path_);
+    ASSERT_TRUE(base::PathExists(audit_results_path_));
+  }
+
+  base::FilePath net_log_path_;
+  base::FilePath audit_results_path_;
+
+  DISALLOW_COPY_AND_ASSIGN(BraveNetworkAuditTest);
+};
+
+// Loads brave://welcome first to simulate a first run and then loads another
+// URL, waiting some time after each load to allow gathering network requests.
+IN_PROC_BROWSER_TEST_F(BraveNetworkAuditTest, BasicTests) {
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), GURL("brave://welcome")));
+  WaitForTimeout(kMaxTimeoutPerLoadedURL);
+
+  ASSERT_TRUE(embedded_test_server()->Start());
+  GURL simple_url(embedded_test_server()->GetURL("/simple.html"));
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), simple_url));
```

Comment:
this is great; i think eventually we should load `brave://rewards`, simulate turning that on, then wait for any more requests. could be done as part of this PR or as a follow up.